sigopt/run_context.py

```diff
@@ -0,0 +1,485 @@
+  precedence = {
+    'failed': 2,
+    'completed': 1,
+  }
+  new_state_precedence = precedence.get(new_state, 0)
+  old_state_precedence = precedence.get(old_state, 0)
+  return new_state_precedence >= old_state_precedence
+
+class RunContext(BaseRunContext):
+  def __init__(
+    self,
+    connection,
+    run,
+    suggestion,
+  ):
+    super().__init__()
+    self.connection = connection
+    self.run = run
+    self.suggestion = suggestion
+    fixed_values = {}
+    if self.suggestion is not None:
+      fixed_values.update(self.suggestion.assignments)
+    self._params = RunParameters(self, fixed_values)
+
+  def to_json(self):
+    data = {"run": self.run.to_json()}
+    if self.suggestion is not None:
+      data["suggestion"] = self.suggestion.to_json()
+    return data
+
+  @classmethod
+  def from_json(cls, data):
+    suggestion = data.get("suggestion")
+    if suggestion is not None:
+      suggestion = Suggestion(suggestion)
+    run = TrainingRun(data["run"])
+    return cls(get_connection(), run, suggestion)
+
+  @property
+  def id(self):
+    return self.run.id
+
+  @property
+  def params(self):
+    return self._params
+
+  def __enter__(self):
+    return self
+
+  def __exit__(self, type, value, tb):
+    self._end(exception=value)
+
+  def _end(self, exception):
+    old_run_state = self.connection.training_runs(self.run.id).fetch().state
+    new_run_state = 'failed' if exception else 'completed'
+    if allow_state_update(new_run_state, old_run_state):
+      self._update_run({'state': new_run_state})
+    if self.suggestion:
+      try:
+        (self.connection.experiments(self.suggestion.experiment)
+          .observations()
+          .create(training_run=self.run.id))
+      except ApiException as e:
+        if e.status_code == 400:
+          self.connection.experiments(self.suggestion.experiment).observations().create(
+            failed=True,
+            training_run=self.run.id,
+          )
+        else:
+          raise
+    print(
+      'Run finished, view it on the SigOpt dashboard at https://app.sigopt.com/run/{run_id}'.format(
+        run_id=self.run.id,
+      )
+    )
+
+  def _request(self, method, path, params, headers=None):
+    base_url = self.connection.impl.api_url
+    run_id = self.run.id
+    return self.connection.impl._request(
+      method=method,
+      url=f'{base_url}/v1/training_runs/{run_id}{path}',
+      params=params,
+      headers=headers,
+    )
+
+  def _update_run(self, body):
+    self._request(
+      method='MERGE',
+      path='',
+      params=body,
+      headers={'X-Response-Content': 'skip'},
+    )
+
+  def _create_checkpoint(self, body):
+    self._request(
+      method='POST',
+      path='/checkpoints',
+      params=body,
+      headers={'X-Response-Content': 'skip'},
+    )
+
+  @updates('assignments')
+  def _set_parameters(self, parameters):
+    return parameters
+
+  @updates('datasets')
+  def _log_dataset(self, name):
+    return {name: {}}
```

Comment:
Why do we log the name with an empty dict value instead of, say, `{'name': name}`?

Reply:
It's the way the API is designed. dataset once had metadata which has been removed, so now its just a dict with empty values. ex. `MERGE /training_runs/X {"mnist": {"cv_split": 0.2, "size": 1024}}`

Reply:
👍